### ISSUE TITLE
Support adding arbitrary HTTP response headers from dependencies

### ISSUE THREAD

**Issue by martincostello** [Microsoft/ApplicationInsights-dotnet-server#587]
It would be beneficial when HTTP dependencies are intercepted to be able to register an interceptor for `HttpResponseMessage` via DI that would allow the integrator to inspect the response's HTTP response headers.

This would allow an integrator to inspect them for values of interest which could be added to the dependency telemetry submitted to Application Insights.

For example, this would allow the activity Id (from `x-ms-activity-id`) and consumed Request Units (from `x-ms-request-charge`) from requests to Azure Cosmos DB to be added to the dependency data ([headers documentation](https://docs.microsoft.com/en-us/rest/api/documentdb/common-documentdb-rest-response-headers)).

For example:

```csharp
public interface IHttpResponseInterceptor
{
    void InspectResponse(HttpResponseMessage response, DependencyTelemetry telemetry)
}

internal class MyHttpResponseInterceptor : IHttpResponseInterceptor
{
    public void InspectResponse(HttpResponseMessage response, DependencyTelemetry telemetry)
    {
        if (response.Headers.TryGetValues("x-ms-activity-id", out IEnumerable<string> values))
        {
            telemetry.Properties["Activity Id"] = string.Join(", ", values);
        }

        if (response.Headers.TryGetValues("x-ms-request-charge", out values))
        {
            telemetry.Properties["RUs"] = string.Join(", ", values);
        }
    }
}

public class Startup
{
    public void ConfigureServices(IServiceCollection services)
    {
        services.AddSingleton<IHttpResponseInterceptor, MyHttpResponseInterceptor>();
    }
}
```

**Comment by SergeyKanzhelev** [comment]
+ @lmolkova to comment

**Comment by lmolkova** [comment]
You can get this with DiagnosticSource: you can subscribe to the particular source depending on the .net flavor you use (.NET Core or Full Framework) and get notifications for each outgoing HTTP request: start and stop.

If you use .NET Full Framework, please refer to the [ApplicationInsights listener](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/HttpDesktopDiagnosticSourceListener.cs) as an reference listener and [subscriber](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/HttpDesktopDiagnosticSourceSubscriber.cs) implementations.

Please note that on .NET Framework, with current version of DiagnositSource package, you get these functionality on .NET 4.6 only, however soon new version will be published that also supports .NET 4.5.

If you use .NET Core 2.0, please refer to this [guide](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/HttpDiagnosticsGuide.md).
And here is ApplicationInsights implementation of [listener](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/HttpCoreDiagnosticSourceListener.cs) and [subscriber](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/HttpCoreDiagnosticSourceListener.cs#L405). It also supports legacy events fired by .net core 1.1 and prior versions, please refer to the comments in the implementation.

More info on DiagnosticSource is [here](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md) and [here](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md).

**Please be careful with subscription and filters. Make sure you always disable application insights URLs, similarly to how [ApplicationInsights SDK does](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/HttpCoreDiagnosticSourceListener.cs#L451)**.

Please let me know if you need more details or have any questions.

**Comment by lmolkova** [comment]
By the way, there is a new [Activity](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/ActivityUserGuide.md) API in .NET. It represents logging context for the operation.
It has `Id` property that uniquely identifies particular request. It also has `RootId` property that identifies operation and common for all requests. 
`Activity.Current` flows with async calls, so you can always use it to get current context.

We also created [protocol](https://github.com/lmolkova/correlation/blob/master/http_protocol_proposal_v1.md) to pass Activity information over HTTP and use `Request-Id` header to pass Id (parent and root information).

**Comment by martincostello** [comment]
Thank you for the notes and details about this, where I've started to flesh out a bar-bones implementation of the use case described in this issue.

However, it requires what I think is an excessively large custom implementation (~350 lines with whitespace and comments) just to intercept a few HTTP response header values, given how much work the Application Insights telemetry module is _already_ doing.

It wouldn't be so bad if some of the internal types in the SDK were public to be able to leverage them (for example the `ApplicationInsightsUrlFilter` class), but I'm having to effectively copy a lot of code out of Application Insights itself to get things working.

The classes required were:
  1. A telemetry module
  1. A diagnostic source listener
  1. A diagnostic source subscriber
  1. A "property fetcher" (which itself contains another two nested private classes)
  1. An Application Insights URL filter

Further, it seems that my custom diagnostic listener runs after [Application Insights collects the properties from the baggage](https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/d9a98dc655552737f844ed0782d13aeef6f7bc89/Src/DependencyCollector/Shared/Implementation/HttpCoreDiagnosticSourceListener.cs#L214-L220) in `Activity.Current`, so any custom values I add are not forwarded on with the dependency telemetry in the context which Application Insights collections, but the values I'm collecting are specific to the dependency call anyway, so wouldn't belong in the properties of the context (`DependencyTelemetry.Context.Properties`) but the properties of the dependency anyway (`DependencyTelemetry.Properties`).

Can some extensibility into the dependency tracking not be made to allow an integrator to add simple amounts of additional custom properties to dependency tracking? This would make this use case **much** easier to implement and test, and much less brittle to changes in implementation details between Application Insights and/or .NET Core.

Another use case this would support is custom labelling if HTTP dependencies (like is already built in for certain Azure services, such as Blobs and Cosmos DB) before they are submitted to Application Insights. For example, labelling the "Dependency type" for a call to `api.twitter.com` as "Twitter" instead of "HTTP", such as "Azure DocumentDB" is shown for Cosmos DB.

![image](https://cloud.githubusercontent.com/assets/1439341/26275433/eed24ddc-3d57-11e7-956d-592882ddc1ac.png)
![image](https://cloud.githubusercontent.com/assets/1439341/26275438/12f386f4-3d58-11e7-82db-4c0268d75d41.png)

As it stands based on the documentation given, it's not possible to actually extend or customise the HTTP dependencies, merely re-implement more-or-less the same functionality and add further information to Application Insights in parallel, rather than _enrich_ the existing telemetry.

A simple DI-enabled extension point for inspecting requests and responses to/from dependencies would enable a lot of more-specific labelling and context for integrators with minimal amounts of custom code to maintain.

**Comment by lmolkova** [comment]
This is a very good feedback, thank you! 

Let me try to summarize it
1. **DiagnosticListener usage**: There should be easy to use and efficient primitives to work with DiagnosticListener, such as generic subscriber, listener and property fetcher. 
There are a couple of packages that could simplify DiagnosticListener usage

   * [System.Reactive.Core](https://github.com/Reactive-Extensions/Rx.NET) package that simplifies subscription process, but does not do payload parsing.

    * [Microsoft.Extensions.DiagnosticAdapter](https://github.com/aspnet/EventNotification) package.  It also takes care for payload extraction. However it does not fully support advanced scenarios with filtering, see https://github.com/aspnet/EventNotification/issues/67, so **I do not recommend using it for now**, we will continue to work with AspNetCore team to improve it. 
It also now targets netstandard1.1 only, however there is an effort to enable all AspNetCore packages on  .NET Framework, see https://github.com/aspnet/Announcements/issues/243.

2. **URL filter**. ApplicationInsights should expose URL filter as public API. I believe there are cases when consumers does not know anything about ApplicationInsights - e.g. another tracing systems and we cannot ask everyone to filter out AppInsights URL. I created a bug to address it #592. 

3. **Changing DependencyTelemetry type**. There is an [existing way](https://docs.microsoft.com/en-us/azure/application-insights/app-insights-api-filtering-sampling)  in AppInsights with custom `ITelemetryIntitializer` implementation. You can override telemetry type if needed there. 

4. **Custom properties**. `Activity` already has `Tags` exactly for this purpose. ApplicationInsights does not respect them for now. I created an issue to track it  https://github.com/Microsoft/ApplicationInsights-dotnet/issues/562.

    * If custom properties are known when request is starting (before it's sent), you can also add Tags to Activity.Current in DiagnosticListener Start (Request) callback and copy them to Telemetry.Properties in TelemtryInitializer (until [tags support](https://github.com/Microsoft/ApplicationInsights-dotnet/issues/562) is implemented). Note that it will only work on .NET Core because of certain limitations in HTTP instrumentation  on .NET 4.5 and 4.6 -  Activity.Current is not accurate in Stop (Response) event in which dependecy is tracked.

   * If custom properties are known only **after** response is received, there is no good way to intercept them with DiangositcSource. Basically the order of listeners is unknown and AppInsights may have already tracked dependency when your listener receives callback. 
There are different approaches to solve it. In case of well-known endpoints such an Azure APIs, we want client libraries to be instrumented (with Activity) so the Activity properties just light up in ApplicationInsights when library is instrumented without any user code. 
However there will always be custom libraries and components that are not instrumented with Activity yet.
@SergeyKanzhelev what are your thoughts on it? Should AppInsights provide API for this assuming Microsoft components will be gradually instrumented?

For the other cases, I think we should leverage existing APIs available through .NET: Activity, DiagnosticListener and its extensions wherever possible and work with partner teams to improve them instead of providing new ApplicationInsights APIs to notify about dependencies.

DiagnosticSource, Activity and its integration with ApplicationInsights is new, so we really appreciate your feedback, based on it we see which scenarios we need to improve.

**Comment by martincostello** [comment]
Thank you for the detailed reply @lmolkova - I'll follow up the links to things you've mentioned when I next get a chance with my [test implementation](https://github.com/martincostello/alexa-london-travel-site/tree/Collect-DocumentDb-Headers/src/LondonTravel.Site/Telemetry) and see how much simpler I can make it with what's already there.

I'm sure ASP.NET Core 2.0 would make it simpler again, but I'm not going to migrate to that until it reaches RTM.

**Comment by SergeyKanzhelev** [comment]
@martincostello I'm glad there is a workaround that works for you.

I believe there are generally three asks:
1. Make correlation ID management easier
2. Allow to extract additional information from the http call (including request and response) and add this information to Application Insights Dependency event
3. Support in-place (not in-channel) filtering of certain dependencies

Cross reference another issues that is related to these asks: 

- "Undistinguishable calls problem" https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/343 
- "Support chunked request and response http dependency tracking" https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/62
- "There should be a kill switch that completely disables Application Insight event collection" https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/590

**Comment by federicobarera** [comment]
@lmolkova 

I have read the documentation relative to the diagnostic source listener. 
The documentation explain how to create logging event (Write) and how to subscribe to those.

What is not clear is, at which stage/when the logging events are pushed (or Written to the DiagnosticSource)
Eg. Your HttpCoreDiagnosticSourceListener.cs listen to the following:

private const string HttpOutEventName = "System.Net.Http.HttpRequestOut";
private const string HttpOutStartEventName = "System.Net.Http.HttpRequestOut.Start";
private const string HttpOutStopEventName = "System.Net.Http.HttpRequestOut.Stop";
private const string HttpExceptionEventName = "System.Net.Http.Exception";
private const string DeprecatedRequestEventName = "System.Net.Http.Request";
private const string DeprecatedResponseEventName = "System.Net.Http.Response";

When those events are pushed?

**Comment by lmolkova** [comment]
@federicobarera  
You may refer to the [HTTP Diagnostics Guide](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/HttpDiagnosticsGuide.md)

The [implementation](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs) is also pretty straightforward.

**Comment by KasthuriAlavala** [comment]
The log path should be able to Correlate Systems using correlation ID. It would be great if we have easy way to associate Correlation Id with dependency telemetry item. Having correlation Id in dependency item is fundamental to me in azure world.

**Comment by lmolkova** [comment]
@KasthuriAlavala 
ApplicationInsights correlates telemetry so if your service receives incoming HTTP request and makes outgoing HTTP request, writes a log or throws an exception, AI tracks it automatically and correlation happens without any code in your application.
Automatic correlation was implemented in AI SDK for Web 2.4/AI SDK for AspNetCore 2.1.

There are some operations that are not instrumented (yet), but you can write a bit of code to track and correlate such operation, [this guide](https://docs.microsoft.com/en-us/azure/application-insights/application-insights-custom-operations-tracking) describes it.

**Comment by KasthuriAlavala** [comment]
My requirement is not automatic correlation. I need to get correlation Id from Dependent response  header  and log in DependencyTelemetry, so that I can give this to external teams to troubleshoot.

Thanks
Kasthuri

**Comment by xt0rted** [comment]
I just ran into a situation where being able to grab a header from an http request and include it in the AI telemetry would have saved me a ton of time.

I'm using [WebJobs.Extensions.SendGrid](https://github.com/Azure/azure-webjobs-sdk-extensions) and the response back includes an `x-message-id` header and I'd like to include that with the dependency call telemetry similar to the Blob & Container values for storage requests. Since this requires a bunch of custom code it was actually easier for me to rewrite my webjob and just dump that out to the TraceWriter than to wire up AI.

I did spend some time trying to get a solution working similar to the one @martincostello linked to. It started to work but it didn't handle all of the scenarios I was hoping to use it for. I think it'd work fine for the SendGrid calls, but I also have some that are triggered inside `ThreadPool.QueueUserWorkItem()` which doesn't seem to play well with `Activity.Current`. So even though I get a copy of the response headers I'm unable to stick them into the static `ConcurrentDictionary` for retrieval later.

**Comment by michaelwstark** [comment]
I've recently ran into a scenario where I would have liked to have been able to specify response headers that I would like tracked against my DependencyTelemetry events.  I have several service partners that request that I send them particular headers from their responses when reporting issues with their APIs before they can begin investigation.  It would be preferable to track these identifiers alongside the failed DependencyTelemetry.

**Comment by lmolkova** [comment]
Re-reading the feedback on this issue, I see there are actually two different problems here
1. back-propagation of the trace context. It may be that correlation id generation/sampling decision happens somewhere downstream and not on the edge border or just additional context is back-propagated in the response. This is something we discussed briefly and do not have a final design on.
@SergeyKanzhelev I think this should become part of the tracing standard eventually. 

2. A broader general purpose feature to enrich a `DependencyTelemetry` based on the request/response. These may apply to the HTTP or other kinds of dependencies. 
The feature does not make sense for RequestTelemetry as user code is executed in the scope of this telemetry, i.e. telemetry (via Activity) could be modified as needed in TelemetryInitializer

Perhaps p2 could become not that relevant in presence of strong, widely adopted distributed tracing standard. But we are not there yet.

**Comment by lmolkova** [comment]
I was thinking about it and one possible approach could be to add raw request/response objects on the DependencyTelemetry - the would exists in the process only and would never be reported to the ingestion endpoint.

Pros:
- small and not invasive change: no new events/handlers/delegates involved, just 2 new properties on the DependencyTelemtery
- full flexibility: in TelemteryIntitializer user can do anything with the raw objects. Read headers, body, anything

Cons:
- Request and Response have to be of `object` type to support SQL and other kinds of dependencies. types have to be checked at runtime by TelemetryIntitializer. 
- some instrumentation mechanisms do not provide access to raw objects such as event source. There might be cases when we are not able to provide HttpResponseMessage (because of .NET Desktop HttpClient limitations) with DiagnosticSource. 

I.e. while it's an easy change from the SDK perspective, users would need to be cautious consuming such raw objects and should expect to receive nothing at all or instance of an unexpected type.

**Comment by michaelwstark** [comment]
@lmolkova.  I am at a point in my iteration where my time is freeing up and I have some cycles to begin contributing towards this work item.  In regards to the approach you documented -- I do like the solution, it seems to provide the most flexibility to the developer to add the details they required to the DependencyTelemetry.  From a selfish perspective, it also aligns with another telemetry platform we leverage in parallel, so I would get to share a good amount of code between the two if I so choose.

It does give someone the ability to shoot them self in the proverbial foot if they choose to log the entire request\response body from a privacy perspective as well as potentially altering their response unintentionally (http response content is a stream that can only be read trivially once IIRC).  That being said I think the former is best decided by the developer, while the later are well-known limitations.

The TelemetryInitializer pattern already requires the author to determine type and cast, so having the Request\Response properties be of type object is natural in my experience with the SDK.

What is the best path forward to settle on a design and begin coding -- I appreciate any advice\guidance you can provide for this first contribution attempt.

**Comment by sandorfr** [comment]
I've implemented some of the things described here in a project recently maybe it could help you @michaelwstark : http://geeklearning.io/application-insights-extensibility-tricks-worth-knowing-1-2/

**Comment by lmolkova** [comment]
@SergeyKanzhelev could you please review the approach we've discussed [above](https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/587#issuecomment-367510081)?

**Comment by lmolkova** [comment]
talked with @SergeyKanzhelev offline. He is ok with the approach.
He also suggested having a list or dictionary of objects attached to the telemetry (similar to [Exception.Data](https://docs.microsoft.com/en-us/dotnet/api/system.exception.data?view=netframework-4.7.1#System_Exception_Data)).

Such change would require
1. [base SDK](http://github.com/Microsoft/ApplicationInsights-dotnet) DependencyTelemetry type change
2. web SDK (this one) change:
   - Http Desktop, Core diagnostic source listeners change
   - Profiler Http instrumentation change
   - SQL diagnostic listener change (optional, but should be useful to demonstrate that approach works for non-http dependencies)

Things to consider: 
- There are some corner cases when we won't be able to get HttpResponseMessage, instead, we'd have response and header collection (see Http Desktop Diagnostic Source Listener) 
- There are Http, SQL and perhaps other types dependencies 
- Memory usage. requests and responses could be huge and we prolong their lifetime. Is there anything we can do about it?


I think the best way to start is to create a quick proof of concept/prototype with some samples/unit test. don't worry about style cop/test coverage at this point. 

If you need any help or have any issues, ping me. We can also have a skype call to go through the details, just let me know.

**Comment by lmolkova** [comment]
This is implemented in #900 

Basically during TelemetryIntitialiers run you have access to HttpRequestMessage/HttpWebRequest and HttpResponseMessage/HttpWebResponse through OperationDetails:

`dependencyTelemetry.TryGetOperationDetail("HttpResponse", out var reponseObject)`

Http keys are:

```csharp
        public const string HttpRequestOperationDetailName = "HttpRequest";
        public const string HttpResponseOperationDetailName = "HttpResponse";
        public const string HttpResponseHeadersOperationDetailName = "HttpResponseHeaders";

```

Be careful: in some cases instead of HttpResponse we can only give you HttpResponseHeaders,
